src/lib.rs

```diff
@@ -0,0 +1,459 @@
+            <T as Trait>::Currency::repatriate_reserved(&from, &to, total_deposit)?;
+
+            Ok(())
+        }
+
+        // Burn the token and unreserve the currency for metadata
+        fn burn(origin, token_id: T::Hash) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+
+            // Ensure token is existed
+            Self::ensure_token_exists(&token_id)?;
+
+            // Check if sender can transfer token
+            Self::ensure_is_approved_or_owner(&token_id, &sender)?;
+
+            // burn the token
+            Self::_burn(&token_id)?;
+
+            // Compute the total deposit fee
+            let total_deposit = Self::get_metadata_fee(&token_id);
+
+            // Repay deposit currency to token owner
+            <T as Trait>::Currency::unreserve(&sender, total_deposit);
+
+            Ok(())
+        }
+
+        // uri not necessary yet
+        // fn register_token_uri(origin, token_id: T::Hash, token_uri: Vec<u8>)-> DispatchResult {
+        //     // Get sender from signature
+        //     let sender = ensure_signed(origin)?;
+
+        //     // Ensure token existed in ERC 721 contract
+        //     Self::ensure_token_exists(&token_id)?;
+
+        //     // Only token's owner can update URI
+        //     Self::ensure_token_owner(&token_id, &sender)?;
+
+        //     // Check the length of token's URI
+        //     Self::ensure_token_uri_valid(&token_uri)?;
+
+        //     // Set token URI
+        //     <RegistryTokenURI<T>>::mutate(token_id, |value| *value = token_uri);
+
+        //     Ok(())
+        // }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    // Get the document root via anchor id
+    fn get_document_root(anchor_id: &T::Hash) -> Result<T::Hash, DispatchError> {
+        match <anchor::Module<T>>::get_anchor_by_id(*anchor_id) {
+            Some(anchor_data) => Ok(anchor_data.doc_root),
+            None => Err(Error::<T>::DocumentNotAnchored.into()),
+        }
+    }
+
+    // Compute metadata fee
+    fn get_metadata_fee(token_id: &T::Hash) -> BalanceOf<T> {
+        // Transfer the deposit after token transfer done
+        let metadata_length: u32 = <RegistryTokenMetadata<T>>::get(token_id).len() as u32;
+
+        Self::compute_metadata_fee(metadata_length)
+    }
+
+    // Compute deposit fee according to length
+    fn compute_metadata_fee(metadata_length: u32) -> BalanceOf<T> {
+        // Deposit for metadata bytes fee
+        let bytes_deposit =
+            <BalanceOf<T>>::from(metadata_length).saturating_mul(T::NFTDepositPerByte::get());
+
+        // Get the fee
+        bytes_deposit.saturating_add(T::NFTDepositBase::get())
+    }
+
+    // ensure sender is uid's validation smart contract
+    fn ensure_sender_is_validation_function(
+        uid: RegistryUid,
+        sender: &T::AccountId,
+    ) -> DispatchResult {
+        Self::ensure_validation_fn_exists(uid)?;
+
+        if <ValidationFn<T>>::get(uid).unwrap() == *sender {
```

Comment:
Ok, I will change it. Actually it is already checked in ensure_validation_fn_exists.

Reply:
It may be checked before, but that check is not thread safe. If you don't want to check it again, please return the validation fn from `Self::ensure_validation_fn_exists(uid)`

Reply:
Accepted. It is an approach to avoid unwrap.